### ISSUE TITLE
[Linux] ctypes packs bitfields Incorrectly

### ISSUE THREAD

**Issue by 92b6f314-a460-4769-a813-17bfff54a4e6** [python/cpython#73939]
BPO | [29753](https://bugs.python.org/issue29753)
--- | :---
Nosy | @terryjreedy, @amauryfa, @jaraco, @abalkin, @meadori, @eryksun, @ztane, @pablogsal, @mleroy003, @miss-islington, @shihai1991, @karlding, @FFY00, @thesamprice
PRs | <li>python/cpython#19850</li><li>python/cpython#27085</li><li>python/cpython#27086</li>
Files | <li>[test.py](https://bugs.python.org/file46708/test.py "Uploaded as text/plain at 2017-03-08.00:36:40 by Charles Machalow"): File to test with</li><li>[ctypes-bitfields-bug.tar.gz](https://bugs.python.org/file47511/ctypes-bitfields-bug.tar.gz "Uploaded as application/x-gzip at 2018-04-01.16:18:14 by @mleroy003"): tar file containing a test program in C and Python 3, plus a makefile. to tets : make rebuild ; make test</li>

<sup>*Note: these values reflect the state of the issue at the time it was migrated and might not reflect the current state.*</sup>

<details><summary>Show more details</summary><p>

GitHub fields:
```python
assignee = None
closed_at = None
created_at = <Date 2017-03-08.00:36:40.502>
labels = ['ctypes', 'type-bug', '3.9', '3.10', '3.11']
title = '[Linux] ctypes packs bitfields Incorrectly'
updated_at = <Date 2021-07-17.02:08:19.325>
user = 'https://bugs.python.org/CharlesMachalow'
```

bugs.python.org fields:
```python
activity = <Date 2021-07-17.02:08:19.325>
actor = 'terry.reedy'
assignee = 'none'
closed = False
closed_date = None
closer = None
components = ['ctypes']
creation = <Date 2017-03-08.00:36:40.502>
creator = 'Charles Machalow'
dependencies = []
files = ['46708', '47511']
hgrepos = []
issue_num = 29753
keywords = ['patch']
message_count = 21.0
messages = ['289193', '289209', '289211', '289212', '301020', '306410', '306434', '306466', '314778', '343998', '363440', '367950', '387821', '392460', '397255', '397257', '397259', '397260', '397261', '397262', '397679']
nosy_count = 15.0
nosy_names = ['terry.reedy', 'amaury.forgeotdarc', 'jaraco', 'belopolsky', 'meador.inge', 'eryksun', 'ztane', 'Charles Machalow', 'pablogsal', 'mleroy003', 'miss-islington', 'shihai1991', 'karlding', 'FFY00', 'thesamprice']
pr_nums = ['19850', '27085', '27086']
priority = 'normal'
resolution = None
stage = 'patch review'
status = 'open'
superseder = None
type = 'behavior'
url = 'https://bugs.python.org/issue29753'
versions = ['Python 3.9', 'Python 3.10', 'Python 3.11']
```

</p></details>

### Linked PRs
* gh-103052

**Comment by 92b6f314-a460-4769-a813-17bfff54a4e6** [comment]
There appears to be a bug related to sizing/packing of ctypes Structures on Linux. I'm not quite sure how, but this structure:

```py
class MyStructure(Structure):
    _pack_      = 1
    _fields_    = [ 
                      ("P",       c_uint16),    # 2 Bytes
                      ("L",       c_uint16, 9),
                      ("Pro",     c_uint16, 1),
                      ("G",       c_uint16, 1),
                      ("IB",      c_uint16, 1),
                      ("IR",      c_uint16, 1),
                      ("R",       c_uint16, 3), # 4 Bytes
                      ("T",       c_uint32, 10),
                      ("C",       c_uint32, 20),
                      ("R2",      c_uint32, 2)  # 8 Bytes
                  ]
```


Gives back a sizeof of 8 on Windows and 10 on Linux. The inconsistency makes it difficult to have code work cross-platform.

Running the given test.py file will print out the size of the structure on your platform.

Tested with Python 2.7.6 and Python 3.4.3 (builtin to Ubuntu 14.04), Python 2.7.13, (built from source) both on Ubuntu 14.04. On Linux all Python builds were 32 bit.

On Windows I tried with 2.7.7 (both 32 and 64 bit).

I believe on both platforms it should return a sizeof 8.

**Comment by 92b6f314-a460-4769-a813-17bfff54a4e6** [comment]
Took a quick look at the c code for this. The area at fault appears to be this section in cfield.c:

```c
#ifndef MS_WIN32
    } else if (bitsize /* this is a bitfield request */
        && *pfield_size /* we have a bitfield open */
        && dict->size * 8 >= *pfield_size
        && (*pbitofs + bitsize) <= dict->size * 8) {
        /* expand bit field */
        fieldtype = EXPAND_BITFIELD;
#endif
```


The problem seems to be after the end of the 2nd c_uint16, it seems to think that the next 10 bytes should extend that c_uint16 to a c_uint32 instead of taking the type as the beginning of a new c_uint32. So I guess it is making the structure something like this:

                      ("P",       c_uint16),
                      ("L",       c_uint32, 9),
                      ("Pro",     c_uint32, 1),
                      ("G",       c_uint32, 1),
                      ("IB",      c_uint32, 1),
                      ("IR",      c_uint32, 1),
                      ("R",       c_uint32, 3),
                      ("T",       c_uint32, 10),
                      # And now this needs an extra 6 bits of padding to fill the c_uint32
                      ("C",       c_uint32, 20),
                      ("R2",      c_uint32, 2)
                      # And now this needs an extra 10 bits of padding to fill the c_uint32.

I guess that is how we get to 10... instead of the expected 8 bytes. I don't believe that this behavior is desired nor really makes logical sense.

**Comment by 92b6f314-a460-4769-a813-17bfff54a4e6** [comment]
Some more debug with print statements in the c code seems to confirm my suspicion:

bitsize, pfield_size, bitofs, dict-\>size prints were added just above the if chain to determine fieldtype and fieldtype was just after that if chain. That code looks something like this:

```c
printf("bitsize: %d\n" , (int)bitsize);
printf("pfield_size: %u\n", (size_t)*pfield_size);
printf("bitofs: %d\n", (int)*pbitofs);
printf("dict->size: %d\n", (int)dict->size);
    if (bitsize /* this is a bitfield request */
        && *pfield_size /* we have a bitfield open */
#ifdef MS_WIN32
        /* MSVC, GCC with -mms-bitfields */
        && dict->size * 8 == *pfield_size
#else
        /* GCC */
        && dict->size * 8 <= *pfield_size
#endif
        && (*pbitofs + bitsize) <= *pfield_size) {
        /* continue bit field */
        fieldtype = CONT_BITFIELD;
#ifndef MS_WIN32
    } else if (bitsize /* this is a bitfield request */
        && *pfield_size /* we have a bitfield open */
        && dict->size * 8 >= *pfield_size
        && (*pbitofs + bitsize) <= dict->size * 8) {
        /* expand bit field */
        fieldtype = EXPAND_BITFIELD;
#endif
    } else if (bitsize) {
        /* start new bitfield */
        fieldtype = NEW_BITFIELD;
        *pbitofs = 0;
        *pfield_size = dict->size * 8;
    } else {
        /* not a bit field */
        fieldtype = NO_BITFIELD;
        *pbitofs = 0;
        *pfield_size = 0;
    }

printf("Fieldtype: %d\n------\n", fieldtype);
```


And the run with the custom-built Python 2.7.13:

```pycon
>>> from test import *
bitsize: 0
pfield_size: 0
bitofs: 255918304
dict->size: 2
Fieldtype: 0
```

------
bitsize: 9
pfield_size: 0
bitofs: 0
dict-\>size: 2
Fieldtype: 1
\------
bitsize: 1
pfield_size: 16
bitofs: 9
dict-\>size: 2
Fieldtype: 2
\------
bitsize: 1
pfield_size: 16
bitofs: 10
dict-\>size: 2
Fieldtype: 2
\------
bitsize: 1
pfield_size: 16
bitofs: 11
dict-\>size: 2
Fieldtype: 2
\------
bitsize: 1
pfield_size: 16
bitofs: 12
dict-\>size: 2
Fieldtype: 2
\------
bitsize: 3
pfield_size: 16
bitofs: 13
dict-\>size: 2
Fieldtype: 2
\------
bitsize: 10
pfield_size: 16
bitofs: 16
dict-\>size: 4
Fieldtype: 3
\------
bitsize: 20
pfield_size: 32
bitofs: 26
dict-\>size: 4
Fieldtype: 1
\------
bitsize: 2
pfield_size: 32
bitofs: 20
dict-\>size: 4
Fieldtype: 2
\------
```pycon
10
>>> MyStructure.P
<Field type=c_ushort, ofs=0, size=2>
>>> MyStructure.T
<Field type=c_uint, ofs=2:16, bits=10>
>>> MyStructure.R
<Field type=c_ushort, ofs=2:13, bits=3>
>>> MyStructure.P
<Field type=c_ushort, ofs=0, size=2>
>>> MyStructure.L
<Field type=c_ushort, ofs=2:0, bits=9>
>>> MyStructure.Pro
<Field type=c_ushort, ofs=2:9, bits=1>
>>> MyStructure.R
<Field type=c_ushort, ofs=2:13, bits=3>
>>> MyStructure.T
<Field type=c_uint, ofs=2:16, bits=10>
>>> MyStructure.C
<Field type=c_uint, ofs=6:0, bits=20>
```

**Comment by eryksun** [comment]
To make it simpler to diagram the fields, I've rewritten your structure using field names A-J:

```py
    import ctypes

    class MyStructure(ctypes.Structure):
        _pack_ = 1
        _fields_ = (('A', ctypes.c_uint16),     # 2 bytes
                    ('B', ctypes.c_uint16, 9),
                    ('C', ctypes.c_uint16, 1),
                    ('D', ctypes.c_uint16, 1),
                    ('E', ctypes.c_uint16, 1),
                    ('F', ctypes.c_uint16, 1),
                    ('G', ctypes.c_uint16, 3),  # 4 bytes
                    ('H', ctypes.c_uint32, 10),
                    ('I', ctypes.c_uint32, 20),
                    ('J', ctypes.c_uint32, 2))  # 8 bytes
```


ctypes is attempting to extend the bitfield for H by switching to a c_uint storage unit with an offset of 2 bytes and adding H field with a 16-bit offset:

```pycon
    >>> MyStructure.H
    <Field type=c_uint, ofs=2:16, bits=10>
```


Here's the correct layout:

    |     uint16    |     uint16    |            uint32             
    |------16-------|------16-------|---10----|--------20---------|-
    A---------------B--------CDEFG--H---------I-------------------J-

and here's the layout that ctypes creates, which wastes 2 bytes:

                    |             uint32            |
    |     uint16    |     uint16    |               |             uint32            
    |------16-------|------16-------|---10----|--6--|--------20---------|-|---10\----
    A---------------B--------CDEFG--H---------------I-------------------J\-----------

The current strategy for extending bitfields to work like gcc on Linux is obviously failing us here. Hopefully someone can flesh out the exact rules to make this code accurate. Bitfields are such a pain.

**Comment by ff59cd45-ebe3-4b3e-9696-65dc59a38b8c** [comment]
To Charles first: "Gives back a sizeof of 8 on Windows and 10 on Linux. The inconsistency makes it difficult to have code work cross-platform." 

The bitfields in particular and ctypes in general have \*never* been meant to be cross-platform - instead they just must need to match the particular C compiler behaviour of the platform, thus the use of these for cross platform work is ill-advised - perhaps you should just use the struct module instead.

However, that said, on Linux, sizeof these structures - packed or not - do not match the output from GCC; unpacked one has sizeof 12 and packed 10 on my Python 3.5, but they're both 8 bytes on GCC. This is a real bug.

GCC says that the bitfield behaviour is: https://gcc.gnu.org/onlinedocs/gcc-4.9.1/gcc/Structures-unions-enumerations-and-bit-fields-implementation.html

Whether a bit-field can straddle a storage-unit boundary (C90 6.5.2.1, C99 and C11 6.7.2.1).

Determined by ABI.
The order of allocation of bit-fields within a unit (C90 6.5.2.1, C99 and C11 6.7.2.1).

Determined by ABI.
The alignment of non-bit-field members of structures (C90 6.5.2.1, C99 and C11 6.7.2.1).

Determined by ABI. 

Thus, the actual behaviour need to be checked from the API documentation of the relevant platform. However - at least for unpacked structs - the x86-64 behaviour is that a bitfield may not cross an addressable unit.

**Comment by 92b6f314-a460-4769-a813-17bfff54a4e6** [comment]
Antti, is there a place in the ctypes documentation that explicitly says ctypes is not meant to be used cross-platform? If not, shouldn't that be mentioned?

I think ultimately ctypes should default to standard OS/compiler behavior, but should allow the flexibility for one to override for cross-platform interchangeability.

In the C++ code here, the code is explicitly checking if Windows (or GCC) to choose behavior. In theory, that could be changed to allow runtime-choice for packing behavior. 

Of course, that is probably best for a feature issue. For this one, I'd just like to have the behavior on Linux match GCC, as that is the definitive bug here as our example has shown.

**Comment by ff59cd45-ebe3-4b3e-9696-65dc59a38b8c** [comment]
"Antti, is there a place in the ctypes documentation that explicitly says ctypes is not meant to be used cross-platform? If not, shouldn't that be mentioned?"

I don't know about that, but the thing is nowhere does it say that it is meant to be used cross-platform. It just says it allows defining C types. It is somewhat implied that C types are not cross-platform at binary level, at all.

**Comment by 92b6f314-a460-4769-a813-17bfff54a4e6** [comment]
All of Python is implicitly cross platform. If something isn't actually cross platform, it should be mentioned explicitly in the documentation. For example see the mmap documentation, it explicitly say on Unix it does X, on Windows it does Y. We should be as explicit as possible for ctypes to say on Windows we expect packing like X (GCC) and on Linux we expect packing like Y (VC++).

**Comment by matthiasgoergens** [comment]
@FFY00 The current code is buggy, but both your fix and @CharlesMachalow misunderstand what `_pack_` is supposed to do.  At least according to the docs.  See https://github.com/python/cpython/pull/97702/files#r990785728

**Comment by matthiasgoergens** [comment]
> I’ve been looking forward to this fix. The old implementation did not match what my LLVM compiler generated.

You might like my fix in https://github.com/python/cpython/pull/97702

**Comment by encukou** [comment]
Fixed in #97702, the reproducer is now part of tests (as the `Example_gh_73939` classe)
